### DISCUSSION TITLE
How to set the type FormData in React-Native (Typescript)?

### DISCUSSION THREAD

**Discussion by yukchit** [react-hook-form/react-hook-form#2436] [Q&A] [answered]
Hi guys, would anyone help me with this question, please? I duno how to deal with react-native's react-hook-form for uploading FormData with image.

**FormScreen:**

  type FormData = {
    image: string;
    title: string;
    description: string;
    created_at: Date;
    latitude: number;
    longitude: number;
  };

or

  type FormData = {
    image: {uri: string, name: "image", type: "image/jpeg"};
    title: string;
    description: string;
    created_at: Date;
    latitude: number;
    longitude: number;
  };

**???**

const onSubmit = ({image, title, description, created_at, latitude, longitude}: FormData) => {
   dispatch(addPhotosThunk(image, title, description, created_at, latitude, longitude)
  };

**Redux-Thunk:**

const formData = new FormData();

export function addPhotosThunk(
  image: string,
  title: string,
  description: string,
  created_at: Date,
  latitude: number,
  longitude: number,
) {
  return async (dispatch: ThunkDispatch, getState: () => RootState) => {
    formData.append("image", {uri: image, name: "image", type: "image/jpg"})
    formData.append("title", title);
    formData.append("description", description);
    formData.append("created_at", created_at);
    formData.append("latitude", latitude);
    formData.append("longitude", longitude);

    const res = await fetch(`${config.BACKEND_URL}/photo/addPhoto`, {
      method: 'POST',
      body: formData, 
      headers: {
        'Content-Type': 'multipart/form-data; ',
      },
    });
    const json = await res.json();
    console.log(json);
    dispatch(addPhotos(json));

    if (res.status == 400) {
      console.log(json.error);
    } else if (res.status == 500) {
      console.log('internal server error');
    }
  };
}

is above correct?

**Comment by bluebill1049** [comment, marked as answer]
i think `image` type should be `FileList`?

**Reply by yukchit** [reply]
typescript seems not to recognize FileList as a type

the react-hook-form in react-native using controller get values from TextInput (so the values are forced to be string), right?
(from the example of [https://react-hook-form.com/get-started#ReactNative](https://react-hook-form.com/get-started#ReactNative) )

and react-native only has TextInput (from [https://reactnative.dev/docs/components-and-apis](https://reactnative.dev/docs/components-and-apis) ), so we need to get the uri of image (after using image-crop-picker package) as a string from TextInput, hv i understood wrong?

**Reply by bluebill1049** [reply]
try with NestedValue: https://react-hook-form.com/ts#NestedValue

**Reply by yukchit** [reply]
> try with NestedValue: https://react-hook-form.com/ts#NestedValue

good inspiration, thanks a lot!


anyway, mine is still okay with following:

**FormScreen:**
type FormData = {
image: string;
title: string;
description: string;
created_at: string;
latitude: string;
longitude: string;
};

(all string because of react-native's TextInput, then can handle these afterwards before appending formData)



**Redux-Thunk:**
```
const formData = new FormData();

export function addPhotosThunk(
image: string,
title: string,
description: string,
created_at: string,
latitude: number,
longitude: number,
) {
  return async (dispatch: ThunkDispatch, getState: () => RootState) => {
    formData.append('image', {
      uri: image,
      name: 'image.jpg',
      type: 'image/jpeg',
    });
    formData.append('title', title);
    formData.append('description', description);
    formData.append('created_at', created_at);
    formData.append('latitude', latitude);
    formData.append('longitude', longitude);
    console.log(`${config.BACKEND_URL}/photo/addPhotoRN/`);
    try {
      const res = await fetch(`${config.BACKEND_URL}/photos/addPhotoRN`, {
        method: 'POST',
        body: formData,
      });
      const json = await res.json();
      console.log('JSON', json);

      if (res.status == 400) {
        console.log(json.error);
      } else if (res.status == 500) {
        console.log('internal server error');
      }
    } catch (err) {
      console.log(err);
    }
  };
}
```

**Reply by bluebill1049** [reply]
awesome.

**Reply by guisszo** [reply]
this saved my life... thanks you 😊
